### PR TITLE
added description and example of `face_dimension`, etc. attributes.

### PR DESCRIPTION
**Pull request by ChrisBarker-NOAA**
This is required to disambiguate row-major and column-major (or fortran of C style, or whatever you want to call it).

This was discussed in this issue in the py_ugrid project:

https://github.com/pyugrid/pyugrid/issues/109

And I _think_ there was consensus to do it this way. And we are in process of implementing it in py_ugrid

NOTE: I'm not totally sure that I added it to ALL the examples appropriately....

Do we need to bump the version number?

@hrajagers, @rsignell-usgs, @bekozi, I'd love your input on this.

### CODE REVIEW

**Comment by rsignell-usgs** [comment]
@ChrisBarker-NOAA , I think you need to add the description of face_dimension to the more generalized case here:
https://github.com/ugrid-conventions/ugrid-conventions/blob/add_face_dimension/src/index.md#2d-flexible-mesh-mixed-triangles-quadrilaterals-etc-topology
modifying the description to apply to meshes with more than 3 nodes

**Comment by ChrisBarker-NOAA** [comment]
will do.

Also, I didn't add it to the 1-D case -- don't think it's relevant there.

File: src/index.md [hunk 1]
```diff
@@ -126,11 +127,15 @@ The topology information is stored as attributes to a dummy variable (in the exa
...
 
+The `face_dimension` attribute specifies with netcdf dimension is used to indicate the index of the face. This is needed because some applications store the data with the fastest varying index first, and some with that index last. The default is to use a (num_faces, 3) array for triangles, but some applications might use a (3, num_faces) order, in which case the `face_dimension` attribute can help the client code disambiguate. The `edge_dimension` attribute is similar for the edge_coordinates array.
```
Comment (rsignell-usgs) [inline src/index.md]: @ChrisBarker-NOAA , how about modifying "specifies with netcdf dimension" to "specified which dimension" ?

**Comment by ChrisBarker-NOAA** [comment]
OK -- added text to the other two cases, and bumped the version number.

Is it time for 1.0 ???

File: src/index.md (hunk 1 above)
Comment (ChrisBarker-NOAA) [inline src/index.md]: On Thu, Feb 18, 2016 at 9:28 AM, Rich Signell <notifications@github.com>
wrote:

> @ChrisBarker-NOAA <https://github.com/ChrisBarker-NOAA> , how about
> modifying "specifies with netcdf dimension" to "specified which dimension" ?
>

duh. Done.

-- 

Christopher Barker, Ph.D.
Oceanographer

Emergency Response Division
NOAA/NOS/OR&R            (206) 526-6959   voice
7600 Sand Point Way NE   (206) 526-6329   fax
Seattle, WA  98115       (206) 526-6317   main reception

Chris.Barker@noaa.gov

**Comment by hrajagers** [comment]
I'm not sure how the GitHub system works exactly with editing the document, so I just write my comments here.

1) The "edge_dimension" is currently listed as an optional attribute. I suggest to move it to the optionally required attributes since if you specify the optionally required attribute "edge_node_connectivity" then it's logical or required to also specify the "edge_dimension". This has already been done for the full 3D case.

2) State explicitly that "edge_dimension" attribute is required if "edge_node_connectivity" is specified. And similarly that "face_dimension" attribute is required if "face_node_connectivity" is specified for the full 3D case.

3) Under "Fully 3D unstructured (i.e. non-layered) mesh topology": remove the = and ; symbols from the attributes table.

**Comment by hrajagers** [comment]
...oops, sorry didn't intend to close it...

**Comment by ChrisBarker-NOAA** [comment]
On Feb 19, 2016, at 7:16 AM, Bert Jagers <notifications@github.com> wrote:

I'm not sure how the GitHub system works exactly with editing the document,
so I just write my comments here.

It's in a branch, so you could go in and edit it directly, but this is
fine, too.

1) The "edge_dimension" is currently listed as an optional attribute. I
suggest to move it to the optionally required attributes since if you
specify the optionally required attribute "edge_node_connectivity" then
it's logical or required to also specify the "edge_dimension".

Do we want to make it required? Or can it be optional, and "C order" be
assumed if it's not there? We've got files in the wild without it, and my
code at least will make that assumption....

Thanks,

CHB

**Comment by GLang-BAW** [comment]
Dear all,

I do not support the idea that attributes like "edge_dimension" shall become required attributes.

How can we (easily) determine e.g. number of edges from the available data:

1. Determine an edge coordinate variable from attribute "edge_coordinates" in the mesh variable.

int Mesh2 ;
	Mesh2:long_name = "UnTRIM Gitternetz, Drei- und Vierecke gemischt, kein SubGrid" ;
	Mesh2:cf_role = "mesh_topology" ;
	Mesh2:topology_dimension = 2 ;
	Mesh2:node_coordinates = "Mesh2_node_lon Mesh2_node_lat" ;
	Mesh2:edge_coordinates = "Mesh2_edge_lon Mesh2_edge_lat" ;
	Mesh2:face_coordinates = "Mesh2_face_lon Mesh2_face_lat" ;
	Mesh2:edge_node_connectivity = "Mesh2_edge_nodes" ;
	Mesh2:edge_face_connectivity = "Mesh2_edge_faces" ;
	Mesh2:face_node_connectivity = "Mesh2_face_nodes" ;
	Mesh2:face_edge_connectivity = "Mesh2_face_edges" ;

2. Get the name of the dimension for the edge coordinate variable and you immediately know the
   number of edges in your grid. Similar for nodes, faces, etc.
   
double Mesh2_edge_lon(nMesh2_edge) ;
	Mesh2_edge_lon:long_name = "geografische Laenge der 2D-Gitter-Kanten, Kantenmitte" ;
	Mesh2_edge_lon:units = "degrees_east" ;
	Mesh2_edge_lon:name_id = 1653 ;
	Mesh2_edge_lon:bounds = "Mesh2_edge_lon_bnd" ;
	Mesh2_edge_lon:standard_name = "longitude" ;

This is a relatively simple two step process. Therefore leave "edge_dimension" to remain an optional attribute.

Günther

**Comment by ChrisBarker-NOAA** [comment]
The edge_dimension attribute does not specify how many edges there are, but
rather to specify _which_ netcdf dimension is the edges , vs nodes. This is
to disambiguate what order the arrays are stored in ( they are 2D arrays ).

I think we could say it's only required if you don't use the standard
order, though.

NOTE: I practice, you can almost always assume that the larger dimension is
the faces/edges, etc dimension, but I think it's better to be explicit
about it.

-CHB

On Feb 19, 2016, at 8:41 AM, Günther <notifications@github.com> wrote:

Dear all,

I do not support the idea that attributes like "edge_dimension" shall
become required attributes.

How can we (easily) determine e.g. number of edges from the available data:

   1. Determine an edge coordinate variable from attribute
   "edge_coordinates" in the mesh variable.

int Mesh2 ;
Mesh2:long_name = "UnTRIM Gitternetz, Drei- und Vierecke gemischt, kein
SubGrid" ;
Mesh2:cf_role = "mesh_topology" ;
Mesh2:topology_dimension = 2 ;
Mesh2:node_coordinates = "Mesh2_node_lon Mesh2_node_lat" ;
Mesh2:edge_coordinates = "Mesh2_edge_lon Mesh2_edge_lat" ;
Mesh2:face_coordinates = "Mesh2_face_lon Mesh2_face_lat" ;
Mesh2:edge_node_connectivity = "Mesh2_edge_nodes" ;
Mesh2:edge_face_connectivity = "Mesh2_edge_faces" ;
Mesh2:face_node_connectivity = "Mesh2_face_nodes" ;
Mesh2:face_edge_connectivity = "Mesh2_face_edges" ;

   1. Get the name of the dimension for the edge coordinate variable and
   you immediately know the number of edges in your grid. Similar for nodes,
   faces, etc.

double Mesh2_edge_lon(nMesh2_edge) ;
Mesh2_edge_lon:long_name = "geografische Laenge der 2D-Gitter-Kanten,
Kantenmitte" ;
Mesh2_edge_lon:units = "degrees_east" ;
Mesh2_edge_lon:name_id = 1653 ;
Mesh2_edge_lon:bounds = "Mesh2_edge_lon_bnd" ;
Mesh2_edge_lon:standard_name = "longitude" ;

This is a relatively simple two step process. Therefore leave
"edge_dimension" to remain an optional attribute.

Günther

—
Reply to this email directly or view it on GitHub
<https://github.com/ugrid-conventions/ugrid-conventions/pull/23#issuecomment-186293463>
.

**Comment by GLang-BAW** [comment]
Hi, well eventually I misunderstand your problem. But if you know the name of the UGRID edge coordinate variable (from the UGRID mesh variable) you immediately know the size (number of edges) as well as the name (or ID) of the dimension which represents "number of edges". You can use name (or ID) to localize the position of this dimension in any array where it is used, independent of the number and the order of dimensions in use with the specific array. Any NetCDF variable (array) holds a tuple of IDs which give you the names (or IDs) of the array dimensions. Günther.

**Comment by hrajagers** [comment]
Dear Günther,

I agree that given the edge (auxiliary) coordinate variables you can
determine the dimension name/id associated with the edges -- i.e. the
edge_dimension as we refer to it now. However, the edge coordinates are
optional, so that route may not be available.

Edges are currently always 2 nodes long (although I could see that
generalized) and hence one could determine the edge_dimension from the
edge_node_connectivity as the dimension unequal to length 2. However, such
an option is not available for thr face_dimension since the number of nodes
per face is more variable for arbitrary polygons.

So, should we write that:

1) Either you use the default dimension order with the
nNodesPerVolume/Face/Edge as fastest dimension, or

2) You specify volume/face/edge_dimension attribute or
volume/face/edge_coordinates attribute and associated variables from which
the dimension can be derived?

Best regards,

Bert

Op 20 feb. 2016 09:37 schreef "Günther" <notifications@github.com>:
>
> Hi, well eventually I misunderstand your problem. But if you know the
name of the UGRID edge coordinate variable (from the UGRID mesh variable)
you immediately know the size (number of edges) as well as the name (or ID)
of the dimension which represents "number of edges". You can use name (or
ID) to localize the position of this dimension in any array where it is
used, independent of the number and the order of dimensions in use with the
specific array. Any NetCDF variable (array) holds a tuple of IDs which give
you the names (or IDs) of the array dimensions. Günther.

**Comment by GLang-BAW** [comment]
Thanks Bert. Now I see the problem. I was a bit fixated to our type of data files. If there are data defined on edges, e.g. normal flow velocity, they always come along with edge coordinates, and in this situation attribute "edge_coordinates" will be also present in the mesh variable. If we have no data defined on edges, we are not interested to know "number of edges" at all. Personally I would prefer your option 1, because this is consistent with the CF metadata concept of cell bounds. Günther.

**Comment by ChrisBarker-NOAA** [comment]
@hrajagers wrote:
> So, should we write that:

> 1) Either you use the default dimension order with the nNodesPerVolume/Face/Edge as fastest dimension,

> or 2) You specify volume/face/edge_dimension attribute or volume/face/edge_coordinates attribute and associated variables from which the dimension can be derived?

1) is good --that will make it semi-optional

2) I don't like so much -- when writing reading (and writing) code, I'd really rather there be "only one way to do it" -- i.e. if you use the non-standard order, you specify the *_dimension attribute. Rather than having to look for that attribute, or look for the coordinates variable, and go through that route, etc.

So I'll put in the text for that, and see what you all think.

**Comment by rsignell-usgs** [comment]
@hrajagers and @ocefpaf  , I'm okay with: https://github.com/ugrid-conventions/ugrid-conventions/blob/f53f86dff14ea54c1e626c3aa4480b4e9ec2a37d/src/index.md , except that I think the version should be v1.0, not v0.10

**Comment by ChrisBarker-NOAA** [comment]
Ping!

Anyone object to merging this and bumping the version to 1.0?

-Chris

**Comment by ocefpaf** [comment]
> Ping!
>
> Anyone object to merging this and bumping the version to 1.0?

+1

**Comment by rsignell-usgs** [comment]
Just waiting to hear back from @ocefpaf, and since he gave the +1, merging.

**Comment by rsignell-usgs** [comment]
@ocefpaf, can you generate the new webpage so that the 1.0 is reflected?
We need to also modify the readme.md to point to the web page instead of actually contain the conventions, right?

**Comment by ocefpaf** [comment]
> @ocefpaf, can you generate the new webpage so that the 1.0 is reflected?
> We need to also modify the readme.md to point to the web page instead of
actually contain the conventions, right?

Will do!

PS: Leaving the US today :-(
But I got a new laptop :-)

**Comment by ocefpaf** [comment]
> Will do!

Done! http://ugrid-conventions.github.io/ugrid-conventions/

**Comment by ChrisBarker-NOAA** [comment]
Looks great, thanks!

**Comment by arthurvd** [comment]
Apologies for bringing this up after the 1.0 bump, but I have a question regarding the (non-)requiredness of the `:face_dimension` attribute. The current standard writes:

> The face_dimension attribute specifies which netcdf dimension is used [..] The default is to use the num_faces as fastest dimension; e.g. a (num_faces, 3) array for triangles, **but some applications might use a (3, num_faces) order, *in which case the face_dimension attribute is required* to help the client code disambiguate.**

I think you agreed that this is a good way (I agree too).
But: only when the default order is not used, the attribute is required. Would that not mean that it should be listed in the  attribute tables under the subheading "optionally required"? Now it is under "required".

**Comment by arthurvd** [comment]
@ChrisBarker-NOAA @hrajagers My previous comment may have stayed unnoticed since the issue was already closed. Would you still have an opinion on my minor point?

**Comment by ChrisBarker-NOAA** [comment]
Sorry about that. Yes, I agree. I just don't have the time to go edit the doc -- if you do a PR, I'll merge it.

**Comment by ChrisBarker-NOAA** [comment]
And if you are afraid it will get lost, you could open an issue for it. this is a merged PR from my changes.

**Pull request merged**